rmw_zenoh_cpp/src/rmw_zenoh.cpp

```diff
@@ -1739,33 +2128,236 @@ rmw_create_service(
+    return nullptr);
+  auto free_request_type_support = rcpputils::make_scope_exit(
+    [service_data, allocator]() {
+      allocator->deallocate(service_data->request_type_support, allocator->state);
+    });
+
+  RMW_TRY_PLACEMENT_NEW(
+    service_data->request_type_support,
+    service_data->request_type_support,
+    return nullptr,
+    RequestTypeSupport, service_members);
+  auto destruct_request_type_support = rcpputils::make_scope_exit(
+    [service_data]() {
+      RMW_TRY_DESTRUCTOR_FROM_WITHIN_FAILURE(
+        service_data->request_type_support->~RequestTypeSupport(),
+        RequestTypeSupport);
+    });
+
+  // Response type support
+  service_data->response_type_support = static_cast<ResponseTypeSupport *>(
+    allocator->allocate(sizeof(ResponseTypeSupport), allocator->state));
+
+  RMW_CHECK_FOR_NULL_WITH_MSG(
+    service_data->response_type_support,
+    "Failed to allocate ResponseTypeSupport",
+    return nullptr);
+  auto free_response_type_support = rcpputils::make_scope_exit(
+    [service_data, allocator]() {
+      allocator->deallocate(service_data->response_type_support, allocator->state);
+    });
+
+  RMW_TRY_PLACEMENT_NEW(
+    service_data->response_type_support,
+    service_data->response_type_support,
+    return nullptr,
+    ResponseTypeSupport, service_members);
+  auto destruct_response_type_support = rcpputils::make_scope_exit(
+    [service_data]() {
+      RMW_TRY_DESTRUCTOR_FROM_WITHIN_FAILURE(
+        service_data->response_type_support->~ResponseTypeSupport(),
+        ResponseTypeSupport);
+    });
+
+  // Populate the rmw_service.
+  rmw_service->data = service_data;
+  rmw_service->implementation_identifier = rmw_zenoh_identifier;
+
+  rmw_service->service_name = rcutils_strdup(service_name, *allocator);
+
+  RMW_CHECK_FOR_NULL_WITH_MSG(
+    rmw_service->service_name,
+    "failed to allocate service name",
+    return nullptr);
+
+  auto free_service_name = rcpputils::make_scope_exit(
+    [rmw_service, allocator]() {
+      allocator->deallocate(const_cast<char *>(rmw_service->service_name), allocator->state);
+    });
+
+  // Zenoh implementation for the service
+
+  // TODO(francocipollone): Replace ros_topic_name_to_zenoh_key by service related function.
+  //                        If this is enough simply rename the method.
+  z_owned_keyexpr_t keyexpr = ros_topic_name_to_zenoh_key(
+    rmw_service->service_name, node->context->actual_domain_id, allocator);
+  auto always_free_ros_keyexpr = rcpputils::make_scope_exit(
+    [&keyexpr]() {
+      z_keyexpr_drop(z_move(keyexpr));
+    });
+  if (!z_keyexpr_check(&keyexpr)) {
+    RMW_SET_ERROR_MSG("unable to create zenoh keyexpr.");
+    return nullptr;
+  }
+
+  RCUTILS_LOG_INFO_NAMED(
+    "rmw_zenoh_cpp", "[rmw_create_service] keyexpr: %s",
+    z_loan(z_keyexpr_to_string(z_loan(keyexpr))));
+
+  z_owned_closure_query_t callback = z_closure(service_data_handler, nullptr, service_data);
```

Comment:
I wonder if we should rely on [fifo channels](https://github.com/eclipse-zenoh/zenoh-c/blob/67428d72935f7576c240ed5ab7170342e3affa17/examples/z_queryable_with_channels.c#L63-L65) to set up the queryable?